### PR TITLE
fix: clean failed challenges

### PR DESCRIPTION
**Pull request by ahmaxed**

### CODE REVIEW

File: client/src/redux/failed-updates-epic.js [hunk 1]
```diff
@@ -19,13 +19,25 @@ import {
...
 
 const key = 'fcc-failed-updates';
 
 function delay(time = 0, fn) {
   return setTimeout(fn, time);
 }
 
+function isSubmitable(failure) {
+  // check if backend challenges have a solution
+  if (
```
Comment (RandellDawson) [inline client/src/redux/failed-updates-epic.js]: More concise way of writing this.
```suggestion
    return failure.payload.challengeType !== backEndProject || failure.payload.solution;
```

File: client/src/redux/failed-updates-epic.js [hunk 2]
```diff
@@ -19,13 +19,25 @@ import {
...
+    failure.payload.challengeType === backEndProject &&
```
Comment (RandellDawson) [inline client/src/redux/failed-updates-epic.js]:
```suggestion
```

File: client/src/redux/failed-updates-epic.js [hunk 3]
```diff
@@ -19,13 +19,25 @@ import {
...
+    !failure.payload.solution
```
Comment (RandellDawson) [inline client/src/redux/failed-updates-epic.js]:
```suggestion
```

File: client/src/redux/failed-updates-epic.js [hunk 4]
```diff
@@ -19,13 +19,25 @@ import {
...
+  ) {
```
Comment (RandellDawson) [inline client/src/redux/failed-updates-epic.js]:
```suggestion
```

File: client/src/redux/failed-updates-epic.js [hunk 5]
```diff
@@ -19,13 +19,25 @@ import {
...
+    return false;
```
Comment (RandellDawson) [inline client/src/redux/failed-updates-epic.js]:
```suggestion
```

File: client/src/redux/failed-updates-epic.js [hunk 6]
```diff
@@ -19,13 +19,25 @@ import {
...
+  }
```
Comment (RandellDawson) [inline client/src/redux/failed-updates-epic.js]:
```suggestion
```

File: client/src/redux/failed-updates-epic.js [hunk 7]
```diff
@@ -19,13 +19,25 @@ import {
...
+  return true;
```
Comment (RandellDawson) [inline client/src/redux/failed-updates-epic.js]:
```suggestion
```

File: client/src/redux/failed-updates-epic.js [hunk 8]
```diff
@@ -19,13 +19,25 @@ import {
...
+}
```
Comment (RandellDawson) [inline client/src/redux/failed-updates-epic.js]:
```suggestion
```

File: client/src/redux/settings/update-legacy-certificate-saga.js [hunk 9]
```diff
@@ -17,7 +17,7 @@ function* updateLegacyCertSaga({
   };
   // shape to update completed challenges in redux store
   let reduxShape = [];
-  for (let obj in challengesToUpdate) {
+  for (var obj in challengesToUpdate) {
```
Comment (RandellDawson) [inline client/src/redux/settings/update-legacy-certificate-saga.js]: Just a question.  Why change this to `var`?

**Review by RandellDawson — CHANGES_REQUESTED** [review]
One suggestion (multi-line not supported yet on GitHub) and one question.

File: client/src/redux/failed-updates-epic.js (hunk 1 above)
Comment (ahmadabdolsaheb) [reply client/src/redux/failed-updates-epic.js]: I was thinking of adding more test for other project types, but I like your version much better.

File: client/src/redux/settings/update-legacy-certificate-saga.js (hunk 9 above)
Comment (ahmadabdolsaheb) [reply client/src/redux/settings/update-legacy-certificate-saga.js]: linter was shouting at me but I will revert this.
Thanks for the review @RandellDawson

File: client/src/redux/failed-updates-epic.js [hunk 10]
```diff
@@ -45,7 +50,15 @@ function failedUpdateEpic(action$, state$) {
     filter(() => store.get(key)),
     filter(() => isOnlineSelector(state$.value)),
     tap(() => {
-      const failures = store.get(key) || [];
+      let failures = store.get(key) || [];
+
+      // delete unsubmitable failed challenges
+      if (!failures.every(isSubmitable)) {
```
Comment (RandellDawson) [inline client/src/redux/failed-updates-epic.js]: I could be wrong but I believe the extra iteration through the failures array (with the `every` method`) is not needed.
```suggestion
      let submitableFailures = failures.filter(isSubmitable);
      if (submitableFailures.length !== failures.length) {
        failures = submitableFailures;
        store.set(key, submitableFailures);
     }
```

File: client/src/redux/failed-updates-epic.js [hunk 11]
```diff
@@ -45,7 +50,15 @@ function failedUpdateEpic(action$, state$) {
...
+        let submitableFailures = failures.filter(isSubmitable);
```
Comment (RandellDawson) [inline client/src/redux/failed-updates-epic.js]: Can be removed if you commit the above change.
```suggestion
```

File: client/src/redux/failed-updates-epic.js [hunk 12]
```diff
@@ -45,7 +50,15 @@ function failedUpdateEpic(action$, state$) {
...
+        failures = submitableFailures;
```
Comment (RandellDawson) [inline client/src/redux/failed-updates-epic.js]: Can be removed if you commit the above change.
```suggestion
```

File: client/src/redux/failed-updates-epic.js [hunk 13]
```diff
@@ -45,7 +50,15 @@ function failedUpdateEpic(action$, state$) {
...
+        store.set(key, submitableFailures);
```
Comment (RandellDawson) [inline client/src/redux/failed-updates-epic.js]: Can be removed if you commit the above change.
```suggestion
```

File: client/src/redux/failed-updates-epic.js [hunk 14]
```diff
@@ -45,7 +50,15 @@ function failedUpdateEpic(action$, state$) {
...
+      }
```
Comment (RandellDawson) [inline client/src/redux/failed-updates-epic.js]: Can be removed if you commit the above change.
```suggestion
```

**Review by RandellDawson — CHANGES_REQUESTED** [review]
@ahmadabdolsaheb See what you think about my suggested changes.

File: client/src/redux/failed-updates-epic.js (hunk 11 above)
Comment (raisedadead) [reply client/src/redux/failed-updates-epic.js]: GitHub now supports multiple line comments in the save review comment: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request :)

Comment (RandellDawson) [reply client/src/redux/failed-updates-epic.js]: I was trying to find how to do that, but I thought it was not implemented yet.  Thanks for the information.

File: client/src/redux/failed-updates-epic.js (hunk 10 above)
Comment (ahmadabdolsaheb) [reply client/src/redux/failed-updates-epic.js]: That is correct. the second iteration is not needed.

**Review by RandellDawson — APPROVED** [review]

**Comment by RandellDawson** [comment]
LGTM.  Are there tests written for this?

**Comment by ahmadabdolsaheb** [comment]
you are right, sagas and epics don't have tests which is causing a lot of inconsistency in the redux store. I would need to do some research on testing those.

for now, would you like me to isolate that function and write a test for it individually?

**Comment by raisedadead** [comment]
Tests please.

**Comment by ahmadabdolsaheb** [comment]
test was just added @raisedadead, let me know if you require further changes.

**Review by raisedadead — APPROVED** [review]

**Pull request merged**